src/app/CommandHandlerImpl.cpp

```diff
@@ -623,20 +625,16 @@ CHIP_ERROR CommandHandlerImpl::FallibleAddStatus(const ConcreteCommandPath & pat
         ChipLogError(DataManagement,
                      "Endpoint=%u Cluster=" ChipLogFormatMEI " Command=" ChipLogFormatMEI " status " ChipLogFormatIMStatus " (%s)",
                      path.mEndpointId, ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mCommandId),
-                     ChipLogValueIMStatus(status), context);
+                     ChipLogValueIMStatus(status.GetStatus()), context);
     }
 
-    return AddStatusInternal(path, StatusIB(status));
-}
-
-CHIP_ERROR CommandHandlerImpl::AddClusterSpecificSuccess(const ConcreteCommandPath & aCommandPath, ClusterStatus aClusterStatus)
-{
-    return AddStatusInternal(aCommandPath, StatusIB(Status::Success, aClusterStatus));
-}
+    if (status.HasClusterSpecificCode())
+    {
+        return AddStatusInternal(
+            path, StatusIB{ (status.IsSuccess() ? Status::Success : Status::Failure), status.GetClusterSpecificCode().Value() });
```

Comment:
This relies on the fact that if cluster specific code exists, then status can only be success/failure.

Are the other states invalid? I.e. can't I have `Status::InvalidAction, 1234` as a `status + cluster_status` pair?
Some comments explaining why this is correct maybe referencing spec items would be helpful. Otherwise I would expect to not lose information and have `StatusIB{ status.GetStatuis(), status.GetClusterSpecificCode().Value())`

when (if?) we convert to std::optional compiler may not ralize that HasClusterSpecifiCode is directly checking for value existence, so maybe:


```cpp
chip::Optional<ClusterStatus> clusterStatus = status.GetClusterSpecificCode();
if (clusterStatus.HasValue()) {
    return AddStatusInternal(
         path, 
         StatusIB{ (status.IsSuccess() ? Status::Success : Status::Failure), 
                    clusterStatus.Value() 
                  });
}
```

Is clearer on intent.